fix(server): regenerate missing person thumbnails

daniele-athome opened the pull request:
Fix #3741.

* Queue face thumbnail generation for person records with empty thumbnailPath (the "forced" version will queue jobs for all people)
* :warning: **Architectural change** :warning: Move face thumbnail generation from face recognition queue to thumbnail generation queue
* Renamed `IPersonRepository.getAll` to `IPersonRepository.getAllForUser` since I needed `getAll` for this PR (very opinionated, I would need guidance from the Immich team to address this change properly)
* Store the assetId used for extracting a person's face in `person.faceAssetId` (it will be reused when regenerating the person's thumbnail, otherwise a random asset will be used)

I moved the job to the thumbnail generation queue because it seemed more appropriate - also given where it's started from. Although the actual job implementation is still inside the facial recognition module.

Next steps would be checking if the thumbnail file is missing and force regeneration (for another PR maybe).

### Tests

- [x] automated tests
- [x] regenerate missing face thumbnails
- [x] regenerate all face thumbnails
- [x] no side effects of empty thumbnail path in explore page (no people showed)
- [x] no side effects of empty thumbnail path in people page (no people showed)
- [x] no side effects of empty thumbnail path in person page (blank missing avatar showed)
- [x] other possible side effects of empty thumbnail path?

daniele-athome committed:
Regenerate missing person thumbnails

bo0tzz commented on server/src/infra/repositories/person.repository.ts:
```diff
@@ -50,7 +50,15 @@ export class PersonRepository implements IPersonRepository {
+  }
+
+  getAllWithoutThumbnail(): Promise<PersonEntity[]> {
+    return this.personRepository.createQueryBuilder('person').where('LENGTH(person.thumbnailPath) = 0').getMany();
```
Other places in the code, such as [asset.repository.ts](https://github.com/immich-app/immich/blob/main/server/src/infra/repositories/asset.repository.ts#L173-L180), use `IsNull()` for the WHERE rather than `LENGTH`.

jrasm91 replied:
You can do = '' instead then

jrasm91 reviewed (commented):
I'd prefer to start storing the asset id on the person. Would you consider making that change here as well?

daniele-athome committed:
Check for empty string instead of zero length

daniele-athome committed:
Remember asset used as person face

jrasm91 commented:
> Done. I kept the random face part as fallback in case the original asset is deleted (and after the first upgrade since the column is new).

That seems fine, but you need `onDelete: 'SET NULL'`, otherwise deleting that asset will fail because of a key violation.

daniele-athome commented:
> That seems fine, but you need `onDelete: 'SET NULL'`, otherwise deleting that asset will fail because of a key violation.

I didn't create a relation - because it's not used anywhere. I did set the cascade in the migration though.

jrasm91 commented:
> > That seems fine, but you need `onDelete: 'SET NULL'`, otherwise deleting that asset will fail because of a key violation.
> 
> I didn't create a relation - because it's not used anywhere. I did set the cascade in the migration though.

Oh I see. The migration looks good. We really want everything to be reflected in the code though. If you deleted all the migrations and started fresh, we should be able to generate the same database structure again by generating a new migration. Since you migration is now out of sync with the person entity, this is no longer the case. The foreign key constraint won't be automatically generated if there isn't a relation in the entity.

daniele-athome committed:
Define entity relation between person and asset via faceAssetId

daniele-athome committed:
Typo

daniele-athome committed:
Fix entity relation

daniele-athome committed:
Tests

daniele-athome committed:
Tests

daniele-athome committed:
Fix code formatting

daniele-athome commented:
I adjusted the existing tests and written some new ones, however the TypeORM check is not passing. I implemented the relation from PersonEntity to AssetEntity, am I missing something? Should I define also the inverse relation maybe?

daniele-athome committed:
Fix import path

daniele-athome committed:
Fix migration

daniele-athome commented:
I have adjusted failing tests and added some new ones.

However there is still that TypeORM job failing... I did refactor the migration to make it more "compliant° with automatic generation (i.e. more granular), but it won't work.

daniele-athome commented:
I can even reproduce it locally. It always generates this:

```typescript
import { MigrationInterface, QueryRunner } from "typeorm";

export class TestMigration1693856515281 implements MigrationInterface {
    name = 'TestMigration1693856515281'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "person" DROP CONSTRAINT "FK_ab768006fba24450b92ead70700"`);
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "person" ADD CONSTRAINT "FK_ab768006fba24450b92ead70700" FOREIGN KEY ("faceAssetId") REFERENCES "assets"("id") ON DELETE SET NULL ON UPDATE NO ACTION`);
    }

}
```

while the constraint is present in the database (all migrations were applied)!!

daniele-athome commented:
Rebased on latest main, but the TypeORM problem remains. I don't know, maybe it's a TypeORM bug:

https://github.com/typeorm/typeorm/issues/5117
https://github.com/typeorm/typeorm/issues/9248
https://github.com/typeorm/typeorm/issues/9256

Or maybe I don't know it enough. Anyway that's it for today.

alextran1502 committed:
format

alextran1502 commented:
@daniele-athome The migration issue is because the migration from the code is different from what you have. I tried to run the migration from the `main`, and this is what gets generated. The foreign key relationship doesn't get applied. I suspect it might have to be applied on the other side of the relation.

```typescript
import { MigrationInterface, QueryRunner } from "typeorm";

export class AddPersonFaceAssetId1693871813999 implements MigrationInterface {
    name = 'AddPersonFaceAssetId1693871813999'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "person" ADD "faceAssetId" uuid`);
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "person" DROP COLUMN "faceAssetId"`);
    }

}

```

daniele-athome committed:
Fix entity and migration

daniele-athome commented:
I think I fixed it. Could you run the checks again please? There was some weird system error on GitHub Actions I believe.

daniele-athome commented:
Nope.

```
Error: Can't use 'tar -xzf' extract archive file: /home/runner/work/_actions/_temp_ec943280-3bca-4ca7-b4f4-d73f7f741161/7e33188b-1840-4004-91e3-18a8fa46b7e4.tar.gz. return code: 2.
```

Maybe once more?  :sweat_smile:

daniele-athome committed:
Linting

daniele-athome commented:
Finally! All done.

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into pr/daniele-athome/3970

jrasm91 commented on server/src/domain/media/media.service.ts:
```diff
@@ -49,6 +52,32 @@ export class MediaService {
       }
     }
 
+    const personList = force
```
Can we rename `personList` to `people`? That is what we use everywhere else.

jrasm91 commented on server/src/domain/media/media.service.ts:
```diff
@@ -49,6 +52,32 @@ export class MediaService {
+            x2: face.boundingBoxX2,
+            y1: face.boundingBoxY1,
+            y2: face.boundingBoxY2,
+          } as BoundingBox,
```
Why is this cast necessary? Also, can you just inline the job data into the `queue` call instead?

jrasm91 commented on server/src/infra/entities/person.entity.ts:
```diff
@@ -36,6 +37,12 @@ export class PersonEntity {
   @Column({ default: '' })
   thumbnailPath!: string;
 
+  @Column({ type: 'uuid', nullable: true })
+  faceAssetId!: string | null;
+
+  @ManyToOne(() => AssetEntity, { onDelete: 'SET NULL', nullable: true })
+  faceAsset!: AssetEntity | null;
+
```
I don't think you need nullable in both column decorators. Also, I don't think you need `type: 'uuid'` in the ID column either. Can you compare this with other relations and make it match?

daniele-athome replied:
Nullable is set in the album entity:

```
@ManyToOne(() => AssetEntity, { nullable: true, onDelete: 'SET NULL', onUpdate: 'CASCADE' })
albumThumbnailAsset!: AssetEntity | null;

@Column({ comment: 'Asset ID to be used as thumbnail', nullable: true })
albumThumbnailAssetId!: string | null;
```

I set the `uuid` type because the server complained about something (sorry, I didn't keep the logs), but that doesn't seem to happen anymore.

jrasm91 replied:
Sounds good. Looks like that is the only other nullable relation we have. 

Right, typeorm would have defaulted to varchar before you added the relation, which obviously would conflict with the foreign key constraint.

jrasm91 commented on server/src/infra/repositories/person.repository.ts:
```diff
@@ -50,7 +50,15 @@ export class PersonRepository implements IPersonRepository {
+  }
+
+  getAllWithoutThumbnail(): Promise<PersonEntity[]> {
+    return this.personRepository.createQueryBuilder('person').where("person.thumbnailPath = ''").getMany();
```
I think this would work as well? We prefer to avoid the query builder where possible.

```suggestion
    return this.personRepository.findBy({ thumbnailPath: "''" });
```

jrasm91 commented on server/src/infra/repositories/person.repository.ts:
```diff
@@ -118,4 +126,13 @@ export class PersonRepository implements IPersonRepository {
+    return this.assetFaceRepository
+      .createQueryBuilder('assetFace')
+      .where('assetFace.personId = :personId', { personId: personId })
+      .orderBy('random()')
```
Does this actually need to be random? Can we just grab the first face? It looks like using `random()` forces postgres to do a full table scan.

jrasm91 replied:
This whole thing can probably be rewritten as just 


```typescript
return this.assetFaceRepository.findByOne( { personId });
```

jrasm91 replied:
This whole thing can probably be rewritten as just 


```typescript
return this.assetFaceRepository.findByOne({ personId });
```

jrasm91 commented on server/src/infra/repositories/person.repository.ts:
```diff
@@ -118,4 +126,13 @@ export class PersonRepository implements IPersonRepository {
+  async getRandomFace(personId: string): Promise<AssetFaceEntity | null> {
+    return this.assetFaceRepository
+      .createQueryBuilder('assetFace')
+      .where('assetFace.personId = :personId', { personId: personId })
```
```suggestion
      .where('assetFace.personId = :personId', { personId })
```

daniele-athome committed:
Remove unneeded cast

daniele-athome committed:
Conventions

daniele-athome committed:
Simplify queries

daniele-athome committed:
Simplify queries

daniele-athome committed:
Remove unneeded typings from entity

jrasm91 commented on server/src/domain/media/media.service.ts:
```diff
@@ -49,6 +52,32 @@ export class MediaService {
+            },
+            assetId: face.assetId,
+            personId: person.id,
+          } as IFaceThumbnailJob,
```
this one as well

daniele-athome replied:
You're right! I just saw that every combination of queue and parameters are typed.

jrasm91 reviewed (approved):
Looks good!

daniele-athome committed:
Remove unneeded cast
